### ISSUE TITLE
Cannot [onResponseReceived] when in state [WaitingForEndOfRequestEntity] when connecting over HTTPS

### ISSUE THREAD

**Issue by ochrons** [akka/akka-http#1930]
I got this error when connecting with `Http().singleRequest` to our backend which is behind https AWS ELB, but only when the backend service was down (so ELB rejected the connection pretty much immediately)

```
[info] 17:55:01.163 DEBUG [akka.io.TcpOutgoingConnection:88] Attempting connection to [api-xyz.com/13.33.244.105:443]
[info] 17:55:01.177 DEBUG [akka.io.TcpOutgoingConnection:88] Connection established to [api-xyz.com:443]
[info] 17:55:02.117 ERROR [akka.http.impl.engine.client.PoolGateway:69] [0 (WaitingForEndOfRequestEntity)] Slot execution failed. That's probably a bug. Please file a bug at https://github.com/akka/akka-http/issues. Slot is restarted.
[info] java.lang.IllegalStateException: Cannot [onResponseReceived] when in state [WaitingForEndOfRequestEntity]
[info]  at akka.http.impl.engine.client.pool.SlotState.illegalState(SlotState.scala:84)
[info]  at akka.http.impl.engine.client.pool.SlotState.onResponseReceived(SlotState.scala:61)
...
[info] 17:55:02.118 WARN  [akka.http.impl.engine.client.PoolGateway:77] [0 (WaitingForEndOfRequestEntity)] Ongoing request [POST /api/v1/projects/6/models/9/files/upload Chunked] was dropped because pool is shutting down
```

The request is a (chunked) file upload from an in-memory Akka stream with gzip compression.

akka-http 10.1.0
akka 2.5.11
scala 2.12.4
java 1.8.0_121
On Windows 10 Pro 64bit

**Comment by jrudolph** [comment]
It means that the server sent back a response before the request entity was even fully sent. We currently don't support this case but we should make sure to handle it more gracefully at least. See also this previous PR: https://github.com/akka/akka-http/pull/1746. I guess if it is a common enough case, we might have to bite the bullet and add the extra complexity to handle it.

**Comment by ochrons** [comment]
The result was that slot was "restarted", connection established again but then nothing. The request didn't complete with an error but the timed out after 60 seconds. This is of course pretty bad behavior for such a failure case.

I would say this would be quite "common" since it happens with AWS load balancers when the service is down. Of course it probably only affects requests with a larger payload, like file uploads etc.

**Comment by vqngoacc** [comment]
According to [rfc7230](https://tools.ietf.org/html/rfc7230#page-55)
> A client sending a message body SHOULD monitor the network connection
   for an error response while it is transmitting the request.  If the
   client sees a response that indicates the server does not wish to
   receive the message body and is closing the connection, the client
   SHOULD immediately cease transmitting the body and close its side of
   the connection.

We're running into this issue as well.  As stated previously we're unable to recover and must wait for a timeout. +1 for at least handling it more gracefully pending a decision for full support

**Comment by deanjarnold** [comment]
OUt team has been following this thread pretty closely the last few months. Our company product teams make extensive use of Akka HTTP and are great advocates of the API and Akka platform.

One of our team's product is experiencing this issue when it sends a large payload request to a remote service using the Akka HTTP Client implementation. The remote service (not our own) rejects the request based on invalid authentication credentials (headers). As the remote service streams the request, it leaves the majority of the payload unread, causing the error described in this issue. The remote service wants to avoid reading the entire large payload request, as it already knows the request credentials are invalid. I believe this is the intention behind the HTTP 1.1 RFC723.

From an Akka HTTP implementation perspective, a mechanism is required to capture the error and recover gracefully. The current implementation ends up with a non-recoverable situation.

We've applied the PR attached to #1746 and it appears to resolve the issue. However, we are uneasy about putting this into production without full knowledge of the potential implications/fallout.

@jrudolph Do you have a timeframe to deliver this behavior?

**Comment by johanandren** [comment]
We are not actively working on this issue right now, so there is no planned fix date. PRs are, as always, welcome.

**Comment by johanandren** [comment]
Unless we should in fact revive #1746, then no PR is needed :)
@raboof WDYT?

**Comment by vqngoacc** [comment]
@raboof Regarding the recommendation to revert back to the legacy pool implementation, we're see the below exception thrown for this scenario, and it introduces some issues for us as we cleanup resources tied to the request.  Though the exception does occur outside our main thread of execution and we're able to process the response fully.  Perhaps the client is still trying to send the request but the source has been terminated.  

`2018-05-10 10:02:25.015-0700 ERROR [galahad-HttpService-akka.actor.default-dispatcher-49] [a.a.ActorSystemImpl] - Outgoing request stream error
java.io.IOException: Reactive stream is terminated, no reads are possible
	at akka.stream.impl.io.InputStreamAdapter.subscriberClosedException(InputStreamSinkStage.scala:117) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.impl.io.InputStreamAdapter.executeIfNotClosed(InputStreamSinkStage.scala:125) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:144) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:135) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.impl.io.InputStreamPublisher.readAndEmit(InputStreamPublisher.scala:56) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.impl.io.InputStreamPublisher.readAndSignal(InputStreamPublisher.scala:50) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.impl.io.InputStreamPublisher$$anonfun$receive$1.applyOrElse(InputStreamPublisher.scala:43) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.actor.Actor.aroundReceive(Actor.scala:517) ~[akka-actor_2.12-2.5.11.jar:2.5.11]
	at akka.actor.Actor.aroundReceive$(Actor.scala:515) ~[akka-actor_2.12-2.5.11.jar:2.5.11]
	at akka.stream.impl.io.InputStreamPublisher.akka$stream$actor$ActorPublisher$$super$aroundReceive(InputStreamPublisher.scala:31) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.actor.ActorPublisher.aroundReceive(ActorPublisher.scala:286) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.actor.ActorPublisher.aroundReceive$(ActorPublisher.scala:272) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.stream.impl.io.InputStreamPublisher.aroundReceive(InputStreamPublisher.scala:31) ~[akka-stream_2.12-2.5.11.jar:2.5.11]
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:590) ~[akka-actor_2.12-2.5.11.jar:2.5.11]
	at akka.actor.ActorCell.invoke_aroundBody0(ActorCell.scala:559) ~[akka-actor_2.12-2.5.11.jar:2.5.11]
	at akka.actor.ActorCell$AjcClosure1.run(ActorCell.scala:1) ~[akka-actor_2.12-2.5.11.jar:2.5.11]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjweaver-1.8.10.jar:1.8.10]
	at akka.monitor.instrumentation.ActorMonitors$$anon$1.processMessage(ActorMonitor.scala:72) ~[prometheus-akka_2.12-0.7.0.jar:0.7.0]
	at akka.monitor.instrumentation.ActorCellInstrumentation.aroundBehaviourInvoke(ActorInstrumentation.scala:61) ~[prometheus-akka_2.12-0.7.0.jar:0.7.0]
	at akka.actor.ActorCell.invoke(ActorCell.scala:552) ~[akka-actor_2.12-2.5.11.jar:2.5.11]
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257) ~[akka-actor_2.12-2.5.11.jar:2.5.11]
	at akka.dispatch.Mailbox.run(Mailbox.scala:224) ~[akka-actor_2.12-2.5.11.jar:2.5.11]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[?:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[?:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_131]`

**Comment by adamw** [comment]
If it's of any help, we're seeing the same error - but only sometimes, and only on Travis, not locally - when running the akka-http sttp tests. The test that fails (if it fails, it's always the same one) uploads a file to a locally running server (so no https), which translates to sending a `HttpEntity.fromPath(ct, b)`.

An example of the failed log is here: https://travis-ci.org/softwaremill/sttp/jobs/380175861

We started getting this failure around 4 months ago, after upgrading to 10.0.11 - but maybe before we were just lucky, who knows :)

**Comment by adamw** [comment]
Btw. the server is also akka-http and all it does is mirrors whatever it receives:
```
post {
          parameterMap { params =>
            entity(as[String]) { body: String =>
              complete(List("POST", "/echo", paramsToString(params), body)
                .filter(_.nonEmpty)
                .mkString(" "))
            }
          }
        }
```

**Comment by raboof** [comment]
@adamw thanks for sharing! 10.0.11 was the version in which the new client pool was introduced, so that checks out. It does sound rather surprising, since `entity(as[String])` should consume the whole request body before responding. Let's keep an eye on this!

**Comment by raboof** [comment]
@ochrons this might now be resolved since #1746 has been merged - would you have any chance to take that for a spin?

**Comment by raboof** [comment]
For now assuming #1746 resolved this issue. Feel free to re-open if you encounter it on 10.1.3 or later!
